### DISCUSSION TITLE
dhcp plugin "DHCP DISCOVER timed out"

### DISCUSSION THREAD

**Discussion by 1377382065** [strongswan/strongswan#2216] [Q&A]
in centos7 , in compile and install latest release version 5.9.14
```
ipsec --version 
Linux strongSwan U5.9.14/K5.4.271-1.el7.elrepo.x86_64
University of Applied Sciences Rapperswil, Switzerland
```

## dhcp plugin config
```
[root@router strongswan]# cat /etc/strongswan/strongswan.d/charon/dhcp.conf 
dhcp {

    # Always use the configured server address.
    # force_server_address = no

    # Derive user-defined MAC address from hash of IKE identity and send client
    # identity DHCP option.
    # identity_lease = no

    # Interface name the plugin uses for address allocation.
    # interface = eth0

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    # load = yes

    # DHCP server unicast or broadcast IP address.
    # server = 127.0.0.1

    # Use the DHCP server port (67) as source port when a unicast server address
    # is configured.
    # use_server_port = yes

    force_server_address = yes
    identity_lease = yes
    interface = ens192
    load = yes
    server = 10.0.0.255
}
```

## ifconfig
ens192: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.0.1  netmask 255.255.255.0  broadcast 10.0.0.255
        inet6 fe80::d361:7fb8:f414:c6a  prefixlen 64  scopeid 0x20<link>
        RX packets 211330  bytes 137953546 (131.5 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 132560  bytes 13641292 (13.0 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 63  base 0x9000  

## dnsmasq config 
dhcp-range=10.0.0.8,10.0.0.254,12h

## try connect to server, i see to error log
11[IKE] peer requested virtual IP %any
11[CFG] sending DHCP DISCOVER for xxxxxx to 10.0.0.255
11[CFG] sending DHCP DISCOVER for xxxxxx to 10.0.0.255
13[MGR] ignoring request with ID 5, already processing
11[CFG] sending DHCP DISCOVER for xxxxxx to 10.0.0.255
15[MGR] ignoring request with ID 5, already processing
11[CFG] sending DHCP DISCOVER for xxxxxx to 10.0.0.255
06[MGR] ignoring request with ID 5, already processing
11[CFG] sending DHCP DISCOVER for xxxxxx to 10.0.0.255
11[CFG] DHCP DISCOVER timed out

## dnsmasq log 
dnsmasq-dhcp[27388]: DHCPOFFER(ens192) 10.0.0.32 xxxxxx
dnsmasq-dhcp[27388]: DHCPDISCOVER(ens192) xxxxxx
dnsmasq-dhcp[27388]: DHCPOFFER(ens192) 10.0.0.32 xxxxxx
dnsmasq-dhcp[27388]: DHCPDISCOVER(ens192) xxxxxx
dnsmasq-dhcp[27388]: DHCPOFFER(ens192) 10.0.0.32 xxxxxx
dnsmasq-dhcp[27388]: DHCPDISCOVER(ens192) xxxxxx
.....

**Comment by tobiasbrunner** [comment]
Could be a firewall issue. It might also be necessary to experiment a bit with the plugin settings (interface vs. no interface, broadcast vs. unicast server address etc.).

**Reply by 1377382065** [reply]
i stop iptables-service and try again , has same question
but same dhcp config can work in version 5.7.2

**Reply by tobiasbrunner** [reply]
> i stop iptables-service

What does that do?



> but same dhcp config can work in version 5.7.2

Possible that the changes in 3c8887326aecc29db332e7d63a1103e08ff64741 had an effect. Again, you might have to play around with the settings.

**Reply by kjonca** [reply]
I have openwrt router, with strongswan  5.9.11-2 (I do not know if "-2" is openwrt part or not)
`# cat /etc/strongswan.d/charon/dhcp.conf 
dhcp {

    # Always use the configured server address.
    # force_server_address = no

    # Derive user-defined MAC address from hash of IKE identity and send client
    # identity DHCP option.
    identity_lease = yes

    # Interface name the plugin uses for address allocation.
    # interface =

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    load = yes

    # DHCP server unicast or broadcast IP address.
    server = 192.168.29.102

    # Use the DHCP server port (67) as source port when a unicast server address
    # is configured.
    # use_server_port = no

}
`
Works without problems. 

With upgrade to:  5.9.14-r4 - it stopped working 

I tried to play with "interface" or force_server_address , changed server addres to broadcast - no luck.

**Reply by 1377382065** [reply]
now i use 5.7.2 , is work fine
```
[root@router ~]# strongswan --version
Linux strongSwan U5.7.2/K5.4.271-1.el7.elrepo.x86_64
University of Applied Sciences Rapperswil, Switzerland

[root@router charon]# cat dhcp.conf 
dhcp {

    # Always use the configured server address.
    # force_server_address = no

    # Derive user-defined MAC address from hash of IKE identity and send client
    # identity DHCP option.
    # identity_lease = no

    # Interface name the plugin uses for address allocation.
    # interface = eth0

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    # load = yes

    # DHCP server unicast or broadcast IP address.
    # server = 127.0.0.1

    # Use the DHCP server port (67) as source port when a unicast server address
    # is configured.
    # use_server_port = yes

    force_server_address = yes
    identity_lease = yes
    interface = ens192
    load = yes
    server = 10.0.0.1

}
```

**Reply by kjonca** [reply]
with configuration
```
interface = br-lan
#server = 192.168.29.102 (i.e. commented)

```
tcpdump showed me:
```
tcpdump -n  -i br-lan port 67 
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on br-lan, link-type EN10MB (Ethernet), snapshot length 262144 bytes
00:48:16.063749 IP 192.168.29.102.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 7a:a7:f5:ba:84:a4, length 353
00:48:16.066917 IP 192.168.29.102.67 > 255.255.255.255.68: BOOTP/DHCP, Reply, length 300
00:48:17.064420 IP 192.168.29.102.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 7a:a7:f5:ba:84:a4, length 353
00:48:17.066090 IP 192.168.29.102.67 > 255.255.255.255.68: BOOTP/DHCP, Reply, length 300
00:48:19.065078 IP 192.168.29.102.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 7a:a7:f5:ba:84:a4, length 353
00:48:19.069378 IP 192.168.29.102.67 > 255.255.255.255.68: BOOTP/DHCP, Reply, length 300
00:48:22.065726 IP 192.168.29.102.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 7a:a7:f5:ba:84:a4, length 353

```
With any other config tcpdump does not show any traffic to/from port 67

**Reply by tobiasbrunner** [reply]
> With any other config tcpdump does not show any traffic to/from port 67

Even if you don't configure an interface at all? Do you get any errors in the log (in any configuration)?

**Comment by kjonca** [comment]
another piece to puzzle: debian box.
```
ii  charon-systemd                                         5.9.13-2+b2                         amd64        strongSwan IPsec client, systemd support
ii  libcharon-extauth-plugins                              5.9.13-2+b2                         amd64        strongSwan charon library (extended authentication plugins)
ii  libcharon-extra-plugins                                5.9.13-2+b2                         amd64        strongSwan charon library (extra plugins)
ii  libstrongswan                                          5.9.13-2+b2                         amd64        strongSwan utility and crypto library
ii  libstrongswan-extra-plugins                            5.9.13-2+b2                         amd64        strongSwan utility and crypto library (extra plugins)
ii  libstrongswan-standard-plugins                         5.9.13-2+b2                         amd64        strongSwan utility and crypto library (standard plugins)
ii  strongswan-charon                                      5.9.13-2+b2                         amd64        strongSwan Internet Key Exchange daemon
ii  strongswan-libcharon                                   5.9.13-2+b2                         amd64        strongSwan charon library
ii  strongswan-starter                                     5.9.13-2+b2                         amd64        strongSwan daemon starter and configuration file parser
ii  strongswan-swanctl                                     5.9.13-2+b2                         amd64        strongSwan IPsec client, swanctl command

```
``` 
%cat /etc/strongswan.d/charon/dhcp.conf
dhcp {

    #ways use the configured server address.
#force_server_address = yes

    # Derive user-defined MAC address from hash of IKE identity.
    identity_lease = yes

    # Interface name the plugin uses for address allocation.
    interface = eno1

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    load = yes


    # DHCP server unicast or broadcast IP address.
        server=192.168.x.x
#       use_server_port=no

}
```
this works.
but with the same config 6.0.0-2 does not work (no other changes), and symptoms are the same as on openwrt.
based on miscellaneous version number it seems something happened between 5.9.13 and 5.9.14?

**Reply by tobiasbrunner** [reply]
This will be fixed in 6.0.1 (see https://github.com/strongswan/strongswan/commit/abbf9d28b0032cf80b79bcacea3146a60800a6dd).

**Reply by kjonca** [reply]
after installing 
`ii  strongswan     6.0.1-1      all          IPsec VPN solution metapackage`
it seems to work. The only thing I do not understand is:
when I set 
`interface = eno1`
it not work (eno1 is real interface)
when I coment it out - it works.

**Reply by tobiasbrunner** [reply]
> The only thing I do not understand is:
> when I set
> `interface = eno1`
> it not work (eno1 is real interface)
> when I coment it out - it works.

Correct, if you set that option, packets are sent over that interface. But since the new release replies are also only accepted if received on it (the latter wasn't correctly enforced in previous releases). However, in your setup, the reply is sent over `lo`. So you either have to configure the new [`interface_receive` option](https://docs.strongswan.org/docs/latest/plugins/dhcp.html#_configuration) to `lo` or the empty string (to accept replies on any interface), or you remove `interface` so neither sending, nor receiving packets is restricted to a specific interface.
